geest/core/polylines_per_grid_cell.py

```diff
@@ -1,44 +1,90 @@
 import os
 from qgis.PyQt.QtCore import QVariant
 from qgis.core import (
+    QgsGeometry,
     QgsVectorLayer,
     QgsField,
     QgsSpatialIndex,
     QgsProcessingFeedback,
 )
 import processing
-from .create_grids import GridCreator
-from .extents import Extents
+from .utilities import GridAligner
 
 
 class RasterPolylineGridScore:
-    def __init__(self, country_boundary, pixel_size, output_path, crs, input_polylines):
+    def __init__(
+        self,
+        country_boundary,
+        pixel_size,
+        working_dir,
+        crs,
+        input_polylines,
+        output_path,
+    ):
         self.country_boundary = country_boundary
         self.pixel_size = pixel_size
-        self.output_path = output_path
+        self.working_dir = working_dir
         self.crs = crs
         self.input_polylines = input_polylines
+        self.output_path = output_path
+        # Initialize GridAligner with grid size
+        self.grid_aligner = GridAligner(grid_size=100)
 
     def raster_polyline_grid_score(self):
         """
         Generates a raster based on the number of input points within each grid cell.
         :param country_boundary: Layer defining the country boundary to clip the grid.
         :param cellsize: The size of each grid cell.
-        :param output_path: Path to save the output raster.
         :param crs: The CRS in which the grid and raster will be projected.
         :param input_polylines: Layer of point features to count within each grid cell.
         """
 
-        # Create grid
-        self.h_spacing = 100
-        self.v_spacing = 100
-        create_grid = GridCreator(h_spacing=self.h_spacing, v_spacing=self.v_spacing)
-        output_dir = os.path.join("output")
-        merged_output_path = os.path.join(output_dir, "merged_grid.shp")
-        grid_layer = create_grid.create_grids(
-            self.country_boundary, output_dir, self.crs, merged_output_path
+        output_dir = os.path.dirname(self.output_path)
+
+        # Define output directory and ensure it's created
+        os.makedirs(output_dir, exist_ok=True)
+
+        # Load grid layer from the Geopackage
+        geopackage_path = os.path.join(
+            self.working_dir, "study_area", "study_area.gpkg"
+        )
+        if not os.path.exists(geopackage_path):
+            raise ValueError(f"Geopackage not found at {geopackage_path}.")
+
+        grid_layer = QgsVectorLayer(
+            f"{geopackage_path}|layername=study_area_grid", "merged_grid", "ogr"
+        )
+
+        area_layer = QgsVectorLayer(
+            f"{geopackage_path}|layername=study_area_polygons",
+            "study_area_polygons",
+            "ogr",
         )
-        grid_layer = QgsVectorLayer(merged_output_path, "merged_grid", "ogr")
+
+        geometries = [feature.geometry() for feature in area_layer.getFeatures()]
+
+        # Combine all geometries into one using unaryUnion
+        area_geometry = QgsGeometry.unaryUnion(geometries)
```

Comment:
What is the logic for doing this here - nice to avoid pulling everything into memory like this if we can...